synapse/handlers/ui_auth/checkers.py

```diff
@@ -237,11 +237,65 @@ async def check_auth(self, authdict: dict, clientip: str) -> Any:
         return await self._check_threepid("msisdn", authdict)
 
 
+class RegistrationTokenAuthChecker(UserInteractiveAuthChecker):
+    AUTH_TYPE = LoginType.REGISTRATION_TOKEN
+
+    def __init__(self, hs: "HomeServer"):
+        super().__init__(hs)
+        self._enabled = bool(hs.config.registrations_require_token)
+        self.store = hs.get_datastore()
+
+    def is_enabled(self) -> bool:
+        return self._enabled
+
+    async def check_auth(self, authdict: dict, clientip: str) -> Any:
+        if "token" not in authdict:
+            raise LoginError(400, "Missing token", Codes.MISSING_PARAM)
+        if not isinstance(authdict["token"], str):
+            raise LoginError(400, "Token must be a string", Codes.INVALID_PARAM)
+        if "session" not in authdict:
+            raise SynapseError(400, "Missing UIA session", Codes.MISSING_PARAM)
+        session = authdict["session"]
+
+        # If the LoginType.REGISTRATION_TOKEN stage has already been completed,
+        # return early to avoid incrementing `pending` again.
+        # TODO: check stored token matches provided token?
+        try:
+            stored_token = await self.store.get_ui_auth_session_data(
+                session,
+                "registration_token",
+                None,
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: {}".format(session))
```

Comment:
I get a circular dependency when I do that, because `synapse/handlers/auth.py` (where `AuthHandler` is defined), imports `synapse.handlers.ui_auth.checkers`. Am I missing something?

Reply:
Aha I see, that is a pain. Well, in that case it's probably fine to use the storage methods directly (it'd be tricky to get around a 2-node import cycle).

But best to leave a comment stating that a cyclic dependency prevents us from using `AuthHandler.get_session_data`, so that a later refactor doesn't fall into the same trap :)

Reply:
Actually, a buried import may work here as well (see https://github.com/matrix-org/synapse/pull/10142#discussion_r654625680). As an example elsewhere: https://github.com/matrix-org/synapse/blob/a0cd8ae8cbe14d2821cbe8fd6b011c4ddc729344/synapse/handlers/federation.py#L1531-L1532

Reply:
Yep, that works